contracts/src/CrossChainApplications/NativeTokenBridge/NativeTokenSource.sol

```diff
@@ -0,0 +1,204 @@
+        } else {
+            revert("NativeTokenSource: invalid action");
+        }
+    }
+
+    /**
+     * @dev See {INativeTokenSource-transferToDestination}.
+     */
+    function transferToDestination(
+        address recipient,
+        TeleporterFeeInfo calldata feeInfo,
+        address[] calldata allowedRelayerAddresses
+    ) external payable nonReentrant {
+        // The recipient cannot be the zero address.
+        require(
+            recipient != address(0),
+            "NativeTokenSource: zero ercipient address"
+        );
+
+        // Lock tokens in this bridge instance. Supports "fee/burn on transfer" ERC20 token
+        // implementations by only bridging the actual balance increase reflected by the call
+        // to transferFrom.
+        uint256 adjustedFeeAmount = 0;
+        if (feeInfo.amount > 0) {
+            adjustedFeeAmount = SafeERC20TransferFrom.safeTransferFrom(
+                IERC20(feeInfo.feeTokenAddress),
+                feeInfo.amount
+            );
+            SafeERC20.safeIncreaseAllowance(
+                IERC20(feeInfo.feeTokenAddress),
+                address(teleporterMessenger),
+                adjustedFeeAmount
+            );
+        }
+
+        uint256 messageID = teleporterMessenger.sendCrossChainMessage(
+            TeleporterMessageInput({
+                destinationBlockchainID: destinationBlockchainID,
+                destinationAddress: nativeTokenDestinationAddress,
+                feeInfo: feeInfo,
+                requiredGasLimit: MINT_NATIVE_TOKENS_REQUIRED_GAS,
+                allowedRelayerAddresses: allowedRelayerAddresses,
+                message: abi.encode(recipient, msg.value)
+            })
+        );
+
+        emit TransferToDestination({
+            sender: msg.sender,
+            recipient: recipient,
+            amount: msg.value,
+            teleporterMessageID: messageID
+        });
+    }
+
+    /**
+     * @dev Unlocks tokens to recipient.
+     */
+    function _unlockTokens(address recipient, uint256 amount) private {
+        require(
+            recipient != address(0),
+            "ERC20TokenSource: zero recipient address"
+        );
+
+        // Transfer to recipient
+        payable(recipient).transfer(amount);
+
+        emit UnlockTokens(recipient, amount);
+    }
+
+    /**
+     * @dev Sends tokens to BURN_FOR_TRANSFER_ADDRESS.
+     */
+    function _burnTokens(uint256 amount) private {
+        payable(BURNED_TX_FEES_ADDRESS).transfer(amount);
+        emit BurnTokens(amount);
+    }
+
+    /**
+     * @dev Update destinationBurnedTotal sent from destination chain
+     * If the new burned total is less than the highest known burned total, this transaction is a no-op.
+     * The burned total on the destination will only ever increase, but new totals may be relayed to this
+     * chain out of order.
+     */
+    function _updateDestinationBurnedTotal(uint256 newBurnTotal) private {
```

Comment:
Nit: Thoughts on renaming this function `_handleBurnTokens` or something similar? The current name doesn't make it clear that the function (potentially) actually burns the tokens itself to me.